File: frontend/src/lib/tracker.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { api } from "@/lib/api";

export type InteractionChannel =
  | "email"
  | "linkedin"
  | "phone"
  | "in_person"
  | "other";
export type InteractionDirection = "inbound" | "outbound";
export type InterviewFormat =
  | "phone_screen"
  | "technical"
  | "system_design"
  | "behavioral"
  | "onsite"
  | "final"
  | "other";
export type InterviewOutcome = "pending" | "pass" | "fail" | "cancelled";

export interface RecruiterInteraction {
  id: string;
  application_id: string;
  user_id: string;
  channel: InteractionChannel;
  direction: InteractionDirection;
  occurred_at: string;
  contact_name: string | null;
  summary: string | null;
  raw_content: string | null;
  created_at: string | null;
}

export interface InterviewEvent {
  id: string;
  application_id: string;
  user_id: string;
  round_label: string;
  format: InterviewFormat;
  scheduled_at: string | null;
  duration_minutes: number | null;
  interviewer_names: string | null;
  interviewer_notes: string | null;
  my_feedback: string | null;
  outcome: InterviewOutcome;
  created_at: string | null;
  updated_at: string | null;
}

export interface FollowUpReminder {
  id: string;
  application_id: string;
  user_id: string;
  due_at: string;
  note: string;
  channel: InteractionChannel | null;
  completed_at: string | null;
  created_at: string | null;
}

export interface ApplicationActivity {
  interactions: RecruiterInteraction[];
  interviews: InterviewEvent[];
  reminders: FollowUpReminder[];
}

// ---- Read ---------------------------------------------------------------

export function useApplicationActivity(applicationId: string | undefined) {
  return useQuery({
    queryKey: ["application-activity", applicationId],
    queryFn: async () => {
      const { data } = await api.get<ApplicationActivity>(
        `/applications/${applicationId}/activity`,
      );
      return data;
    },
    enabled: !!applicationId,
  });
}

export function useOpenReminders() {
  return useQuery({
    queryKey: ["tracker", "open-reminders"],
    queryFn: async () => {
      const { data } = await api.get<FollowUpReminder[]>(`/tracker/reminders`);
      return data;
    },
    staleTime: 1000 * 60,
  });
}

// ---- Recruiter interactions --------------------------------------------

export function useAddInteraction(applicationId: string) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (payload: {
      channel: InteractionChannel;
      direction: InteractionDirection;
      occurred_at: string;
      contact_name?: string | null;
      summary?: string | null;
    }) => {
      const { data } = await api.post<RecruiterInteraction>(
        `/applications/${applicationId}/interactions`,
        payload,
      );
      return data;
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["application-activity", applicationId] });
    },
  });
}

export function useDeleteInteraction(applicationId: string) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (id: string) => {
      await api.delete(`/applications/${applicationId}/interactions/${id}`);
    },
    onSuccess: () =>
      qc.invalidateQueries({ queryKey: ["application-activity", applicationId] }),
  });
}

// ---- Interview events --------------------------------------------------

export function useAddInterview(applicationId: string) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (payload: {
      round_label: string;
      format: InterviewFormat;
      scheduled_at?: string | null;
      duration_minutes?: number | null;
      interviewer_names?: string | null;
    }) => {
      const { data } = await api.post<InterviewEvent>(
        `/applications/${applicationId}/interviews`,
        payload,
      );
      return data;
    },
    onSuccess: () =>
      qc.invalidateQueries({ queryKey: ["application-activity", applicationId] }),
  });
}

export function useUpdateInterview(applicationId: string) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({
      id,
      patch,
    }: {
      id: string;
      patch: Partial<Omit<InterviewEvent, "id" | "application_id" | "user_id" | "created_at" | "updated_at">>;
    }) => {
      const { data } = await api.patch<InterviewEvent>(
        `/applications/${applicationId}/interviews/${id}`,
        patch,
      );
      return data;
    },
    onSuccess: () =>
      qc.invalidateQueries({ queryKey: ["application-activity", applicationId] }),
  });
}

export function useDeleteInterview(applicationId: string) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (id: string) => {
      await api.delete(`/applications/${applicationId}/interviews/${id}`);
    },
    onSuccess: () =>
      qc.invalidateQueries({ queryKey: ["application-activity", applicationId] }),
  });
}

// ---- Follow-up reminders -----------------------------------------------

export function useAddReminder(applicationId: string) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (payload: {
      due_at: string;
      note: string;
      channel?: InteractionChannel | null;
    }) => {
      const { data } = await api.post<FollowUpReminder>(
        `/applications/${applicationId}/reminders`,
        payload,
      );
      return data;
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["application-activity", applicationId] });
      qc.invalidateQueries({ queryKey: ["tracker", "open-reminders"] });
    },
  });
}

export function useCompleteReminder(applicationId: string) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (id: string) => {
      const { data } = await api.post<FollowUpReminder>(
        `/applications/${applicationId}/reminders/${id}/complete`,
      );
      return data;
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["application-activity", applicationId] });
      qc.invalidateQueries({ queryKey: ["tracker", "open-reminders"] });
    },
  });
}

export function useDeleteReminder(applicationId: string) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (id: string) => {
      await api.delete(`/applications/${applicationId}/reminders/${id}`);
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["application-activity", applicationId] });
      qc.invalidateQueries({ queryKey: ["tracker", "open-reminders"] });
    },
  });
}

File: backend/app/infrastructure/db/models/tracker.py
"""Phase H tracker tables — recruiter interactions, interview events,
follow-up reminders.

All three hang off ``applications`` with ``ON DELETE CASCADE`` and carry
a redundant ``user_id`` so we can tenant-scope queries without a join.
"""
from __future__ import annotations

from datetime import datetime
from uuid import UUID, uuid4

from sqlalchemy import (
    DateTime,
    Enum,
    ForeignKey,
    SmallInteger,
    String,
    Text,
    func,
)
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.infrastructure.db.base import Base

INTERACTION_CHANNELS = ("email", "linkedin", "phone", "in_person", "other")
INTERACTION_DIRECTIONS = ("inbound", "outbound")
INTERVIEW_FORMATS = (
    "phone_screen",
    "technical",
    "system_design",
    "behavioral",
    "onsite",
    "final",
    "other",
)
INTERVIEW_OUTCOMES = ("pending", "pass", "fail", "cancelled")


class RecruiterInteraction(Base):
    __tablename__ = "recruiter_interactions"

    id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True), primary_key=True, default=uuid4
    )
    application_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("applications.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    user_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    channel: Mapped[str] = mapped_column(
        Enum(*INTERACTION_CHANNELS, name="interaction_channel", create_type=False),
        nullable=False,
        default="email",
    )
    direction: Mapped[str] = mapped_column(
        Enum(*INTERACTION_DIRECTIONS, name="interaction_direction", create_type=False),
        nullable=False,
    )
    occurred_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )
    contact_name: Mapped[str | None] = mapped_column(String(255), nullable=True)
    summary: Mapped[str | None] = mapped_column(Text, nullable=True)
    raw_content: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )


class InterviewEvent(Base):
    __tablename__ = "interview_events"

    id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True), primary_key=True, default=uuid4
    )
    application_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("applications.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    user_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    round_label: Mapped[str] = mapped_column(String(120), nullable=False)
    format: Mapped[str] = mapped_column(
        Enum(*INTERVIEW_FORMATS, name="interview_format", create_type=False),
        nullable=False,
        default="phone_screen",
    )
    scheduled_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    duration_minutes: Mapped[int | None] = mapped_column(SmallInteger, nullable=True)
    interviewer_names: Mapped[str | None] = mapped_column(Text, nullable=True)
    interviewer_notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    my_feedback: Mapped[str | None] = mapped_column(Text, nullable=True)
    outcome: Mapped[str] = mapped_column(
        Enum(*INTERVIEW_OUTCOMES, name="interview_outcome", create_type=False),
        nullable=False,
        default="pending",
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )


class FollowUpReminder(Base):
    # Partial unique index on `(user_id, due_at) WHERE completed_at IS NULL`
    # is created in migration 0026 — not declared at the model layer because
    # SQLAlchemy's text-based partial index DSL conflicts with hand-written
    # alembic migrations.
    __tablename__ = "follow_up_reminders"

    id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True), primary_key=True, default=uuid4
    )
    application_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("applications.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    user_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    due_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )
    channel: Mapped[str | None] = mapped_column(
        Enum(*INTERACTION_CHANNELS, name="interaction_channel", create_type=False),
        nullable=True,
    )
    note: Mapped[str] = mapped_column(Text, nullable=False)
    completed_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
